kafka-connect-s3/src/test/java/io/confluent/connect/s3/TopicPartitionWriterTest.java

```diff
@@ -646,6 +642,221 @@ public void testWriteRecordTimeBasedPartitionFieldTimestampHours() throws Except
+    ));
+
+    verify(expectedFiles, 2, schema, records);
+  }
+
+  @Test
+  public void testLateDataPartitionRotationRollSmallFile() throws Exception {
+
+    localProps.put(S3SinkConnectorConfig.FLUSH_SIZE_CONFIG, "1000");
+    localProps.put(
+        S3SinkConnectorConfig.ROTATE_INTERVAL_MS_CONFIG,
+        String.valueOf(TimeUnit.MINUTES.toMillis(1))
+    );
+    localProps.put(
+        StorageSinkConnectorConfig.APPEND_LATE_DATA,
+        String.valueOf(false)
+    );
+    setUp();
+
+    // Define the partitioner
+    Partitioner<FieldSchema> partitioner = new HourlyPartitioner<>();
+    parsedConfig.put(PartitionerConfig.TIMESTAMP_EXTRACTOR_CLASS_CONFIG, "RecordField");
+    parsedConfig.put(PartitionerConfig.SCHEMA_GENERATOR_CLASS_CONFIG, TimeBasedSchemaGenerator.class);
+    partitioner.configure(parsedConfig);
+
+    TopicPartitionWriter topicPartitionWriter = new TopicPartitionWriter(
+        TOPIC_PARTITION, storage, writerProvider, partitioner, connectorConfig, context);
+
+    String key = "key";
+    Schema schema = createSchemaWithTimestampField();
+
+    DateTime first = new DateTime(2017, 3, 2, 10, 0, DateTimeZone.forID("America/Los_Angeles"));
+    // One record every 20 sec, puts 3 records every minute/rotate interval
+    long advanceMs = 20000;
+    long t0 = first.getMillis();
+    int size = 3;
+
+    ArrayList<Struct> records = new ArrayList<>(size);
+    // 0 sec
+    records.add(createRecordWithTimestampField(schema, t0));
+
+    // -20 sec
+    long tminus20 = t0- advanceMs;
+    records.add(createRecordWithTimestampField(schema, tminus20));
+
+    // 120 sec: This should flush
+    long tplus120 = t0 + 120000;
+    records.add(createRecordWithTimestampField(schema, tplus120));
+
+    Collection<SinkRecord> sinkRecords = createSinkRecords(records.subList(0, size), key, schema);
+
+    for (SinkRecord record : sinkRecords) {
+      topicPartitionWriter.buffer(record);
+    }
+
+    topicPartitionWriter.write();
+    topicPartitionWriter.close();
+
+
+    List<String> expectedFiles = new ArrayList<>();
+    // Record at 0 sec
+    String encodedPartition = getTimebasedEncodedPartition(t0);
+    String dirPrefix = partitioner.generatePartitionedPath(TOPIC, encodedPartition);
+    expectedFiles.add(FileUtils.fileKeyToCommit(
+        topicsDir, dirPrefix, TOPIC_PARTITION, 0, extension, ZERO_PAD_FMT //TODO
+    ));
+    // Record at -20 sec
+    String earlyEncodedPartition = getTimebasedEncodedPartition(tminus20);
+    String earlyDirPrefix = partitioner.generatePartitionedPath(TOPIC, earlyEncodedPartition);
+    expectedFiles.add(FileUtils.fileKeyToCommit(
+        topicsDir, earlyDirPrefix, TOPIC_PARTITION, 1, extension, ZERO_PAD_FMT
+    ));
+
+
+    Collections.sort(records, new Comparator<Struct>() {
+      @Override
+      public int compare(Struct o1, Struct o2) {
+        long ts1 = (long) o1.get("timestamp");
+        long ts2 = (long) o2.get("timestamp");
+        return Long.compare(ts1, ts2);
```

Comment:
I bet `return Long.compare((long) o1.get("timestamp"), (long) o2.get("timestamp"));` (not crazy about var names such as `ts1`, `ts2`, ...)
For the same reason, in comparators, and other generic methods with two arguments, better names for the args are:
`public int compare(Struct left, Struct right) {` or `public int compare(Struct lhs, Struct rhs) {`